#160754422 Create crud endpoints for food items categories

AwaMelvine opened the pull request:
#### What does this PR do?
Creates all the endpoints for food items categories

#### Description of Task to be completed?
Created the endpoints to:
- Get all food items categories
- Get particular food item category by ID
- Create new food item category 
- Update existing food item category
- Delete food item category

#### How should this be manually tested?
Using postman and making a request to the following endpoints with the data specified:

GET  `/api/v1/categores` to get all categories
POST `/api/v1/categories`  with data `{ id: 1, name: 'Salad', description: 'healthy!' }` to create
GET `/api/v1/categories/1` to get category with ID of 1
PUT `/api/v1/categories/1` with data `{  name: 'Vegetable', description: 'healthy!' }` to update
DELETE  `/api/v1/categories/1` to delete category with ID of 1


#### What are the relevant pivotal tracker stories?
[#160753794] Create CRUD api endpoints for food item categories with tests

AwaMelvine committed:
Avoid use of wildcard import in routes

AwaMelvine committed:
Write endpoints for categories together with their tests

AwaMelvine committed:
Implement validation of categories

robocopkaka commented on api/v1/controllers/categories.js:
```diff
@@ -0,0 +1,53 @@
+import { toInt } from '../helpers/functions';
```
Remember that you're to prefer using `parseInt` directly over using the helper function

AwaMelvine replied:
Okay I'll change it

robocopkaka commented on api/v1/controllers/categories.js:
```diff
@@ -0,0 +1,53 @@
+  getCategoryById(req, res) {
+    const categoryId = toInt(req.params.categoryId);
+    if (!categoryId) {
+      res.status(422).send({ errors: { categoryId: 'Category Id is required' } });
```
Prefer using 400 for bad requests and not 422 which is for unprocessable entity

AwaMelvine replied:
Noted!

robocopkaka commented on api/v1/controllers/categories.js:
```diff
@@ -0,0 +1,53 @@
+    const category = new Category(req.body);
+    allCategories.push(category);
+
+    res.status(200).json(allCategories);
```
Don't know if I gave this feedback yesterday, but for methods that handle POST and PUT, prefer to return the object that was just created or modified in the response, instead of returning the array that has all the items

AwaMelvine replied:
Yeah. You gave the feedback. I'll update it accordingly

robocopkaka commented on api/v1/controllers/categories.js:
```diff
@@ -0,0 +1,53 @@
+    const updatedCategories = allCategories.splice(index, 1, updatedCategory);
+
+
+    Object.assign(allCategories, updatedCategories);
```
You could look up the `spread` operator. It's usually preferred. Though its not urgent

AwaMelvine replied:
Noted!

robocopkaka commented on api/v1/controllers/foodItems.js:
```diff
@@ -3,49 +3,51 @@ import { toInt } from '../helpers/functions';
-};
 
+    Object.assign(allFoodItems, updatedFoodItems);
+    res.status(201).json(updatedFoodItem);
```
PUT requests usually respond with a `200` status code

robocopkaka commented on api/v1/helpers/validateCategories.js:
```diff
@@ -0,0 +1,16 @@
+  }
+
+  if (Object.keys(errors).length !== 0) {
+    return res.status(422).json({ errors });
```
Prefer 400 here

AwaMelvine replied:
Okay.

robocopkaka commented on test/categories.test.js:
```diff
@@ -0,0 +1,166 @@
+const should = chai.should();
+chai.use(chaiHttp);
+
+const initialCategory = new Category({
```
Remember that this data should ideally not be inside the test. You should create a separate file for it and import it here.

AwaMelvine replied:
Okay. I'll do so

robocopkaka commented on test/categories.test.js:
```diff
@@ -0,0 +1,166 @@
+        });
+    });
+  });
+  describe('Category model', () => {
```
Remember that you're to separate controller and model tests

robocopkaka commented on test/categories.test.js:
```diff
@@ -0,0 +1,166 @@
+        .delete(`/api/v1/categories/${categoryId}`)
+        .end((err, res) => {
+          res.should.have.status(204);
+          expect(res.body).to.deep.equal({});
```
Don't think you should use `deep.equal` on an empty object. Prefer `.equal`

robocopkaka commented on test/categories.test.js:
```diff
@@ -0,0 +1,166 @@
+        .end((err, res) => {
+          res.should.have.status(200);
+          expect(res).to.be.a.json;
+          done();
```
Assert that the object you are expecting is actually returned. 
Hint: import the object into this file. Assert that the object in the response body matches the imported object

robocopkaka commented on test/categories.test.js:
```diff
@@ -0,0 +1,166 @@
+        .get('/api/v1/categories')
+        .end((err, res) => {
+          res.should.have.status(200);
+          expect(res.body.length).to.be.equal(1);
```
Also assert that the expected array is returned. Use the same hint as above

robocopkaka commented on test/categories.test.js:
```diff
@@ -0,0 +1,166 @@
+        .get(`/api/v1/categories/${categoryId}`)
+        .end((err, res) => {
+          res.should.have.status(422);
+          expect(res.body).to.have.property('errors');
```
Try to assert the actual error message returned

robocopkaka commented on test/categories.test.js:
```diff
@@ -0,0 +1,166 @@
+        .put(`/api/v1/categories/${categoryId}`)
+        .send(modifiedCategory)
+        .end((err, res) => {
+          res.should.have.status(422);
```
Remember that you're to stick with either `should` or `expect` in a file. Try to go for consistency

AwaMelvine committed:
Pull changes from develop into food-items

Get the food-items branch to be up to date with the
develop branch before you merge create a pull request
from the food-items branch.

AwaMelvine committed:
- use .equal instead of .deep.equal()
- update should return status code 200
- invalid data should return status code 400
- Use spread operator instead of Object.assign()
- test the value of the data being returned from tests

AwaMelvine committed:
Use expect and remove should

Please enter the commit message for your changes. Lines starting

AwaMelvine committed:
Create a folder called data and define all data in it.
This data can then be imported into tests files and used.
Test code for controllers should be in controllers folder
and that for models should be in the models folder under the test
directory.

AwaMelvine committed:
Organize test code for orders and food items such that
controller tests are placed in the controller folder
and model tests are placed in the model folder.
Also, place testing data for both inside the data folder.
